scala/lambdas/preingest-tdr-aggregator/src/main/scala/uk/gov/nationalarchives/eventaggregator/Aggregator.scala

```diff
@@ -0,0 +1,128 @@
+        )
+      )
+    }
+
+    private def startNewGroup(groupCacheRef: Ref[F, Map[String, Group]], sourceId: String, config: Config, lambdaTimeoutTime: MilliSeconds)(using
+        sfnClient: DASFNClient[F],
+        enc: Encoder[SFNArguments]
+    ): F[GroupId] = {
+      val groupExpiryTime = lambdaTimeoutTime + (config.maxSecondaryBatchingWindow * 1000)
+      val waitFor: Seconds = Math.ceil((groupExpiryTime - Generators().generateInstant.toEpochMilli).toDouble / 1000).toInt.seconds
+      val groupId: GroupId = GroupId(config.sourceSystem)
+      val batchId = BatchId(groupId)
+      for {
+        _ <- sfnClient.startExecution(config.sfnArn, SFNArguments(groupId, batchId, waitFor), batchId.batchValue.some)
+        _ <- groupCacheRef.update(groupCache => groupCache + (sourceId -> Group(groupId, Instant.ofEpochMilli(groupExpiryTime), 1)))
+      } yield groupId
+    }
+
+    private def getNewOrExistingGroupId(groupCacheRef: Ref[F, Map[String, Group]], config: Config, sourceId: String, lambdaTimeoutTime: MilliSeconds)(using
+        Encoder[SFNArguments]
+    ): F[GroupId] = {
+      def log = logWithReason(sourceId)
+      groupCacheRef.get.flatMap { groupCache =>
+        val potentialCurrentGroupForSource = groupCache.get(sourceId)
+        potentialCurrentGroupForSource match
+          case None => log(NoExistingGroup) >> startNewGroup(groupCacheRef, sourceId, config, lambdaTimeoutTime)
+          case Some(currentGroupForSource) =>
+            if lambdaTimeoutTime >= currentGroupForSource.expires.toEpochMilli then
+              log(ExpiryBeforeLambdaTimeout) >> startNewGroup(groupCacheRef, sourceId, config, lambdaTimeoutTime)
+            else if currentGroupForSource.itemCount > config.maxBatchSize then log(MaxGroupSizeExceeded) >> startNewGroup(groupCacheRef, sourceId, config, lambdaTimeoutTime)
+            else groupCacheRef.update(_ + (sourceId -> currentGroupForSource.copy(itemCount = currentGroupForSource.itemCount + 1))) >> Async[F].pure(currentGroupForSource.groupId)
+      }
+    }
+
+    override def aggregate(config: Config, ref: Ref[F, Map[String, Group]], messages: List[SQSMessage], remainingTimeInMillis: Int)(using
+        Encoder[SFNArguments],
+        Decoder[Input]
+    ): F[List[Outcome[F, Throwable, Unit]]] = {
+      val lambdaTimeoutTime = (Generators().generateInstant.toEpochMilli + remainingTimeInMillis).milliSeconds
+      for {
+        fibers <- messages.parTraverse { record =>
+          val process = for {
+            groupId <- getNewOrExistingGroupId(ref, config, record.getEventSourceArn, lambdaTimeoutTime)
+            input <- Async[F].fromEither(decode[Input](record.getBody))
+            _ <- writeToLockTable(input, config, groupId)
+          } yield ()
+          process.start
+        }
+        results <- fibers.traverse(_.join)
+        _ <- logger.info(Map("success" -> results.count(_.isSuccess).toString, "failures" -> results.count(_.isError).toString))("Aggregation complete")
```

Comment:
Even though "success" can be plural, do you think it should it be "successes" instead, to clearly indicate that there are multiple?

Reply:
Yeah ok. It matches more explicitly with failures that way as well.